Plans: Plan Details

frosty opened the pull request:
Adds a plan details view, as part of #4745. This is currently just the detail view itself, with no comparison / plan switching functionality. It shows the details for the plan selected from `PlanListViewController`.

![simulator screen shot 5 feb 2016 11 37 00](https://cloud.githubusercontent.com/assets/4780/12845433/bf76f582-cbfd-11e5-9cf4-e1e33a3566b2.png)

### Known issues

– As with #4772, prices are hardcoded – this will be dealt with as part of #4746. The purchase button is also currently shown for the Free plan.
– The current active plan isn't highlighted yet. This will come with the comparison / switcher update.
– The data is currently hardcoded into the app. It would be nice as a future improvement to fetch this from our API instead to keep up to date.
– A few design (colour, mainly) tweaks may be required. iPad is currently not dealt with, as that'll be handled by the switcher view controller that wraps this one.

Needs Review: @koke

koke commented on WordPress/Classes/Models/Plan.swift:
```diff
@@ -106,3 +146,74 @@ extension Plan {
+}
+
+func ==(lhs: PlanFeature, rhs: PlanFeature) -> Bool {
+    return lhs.title == rhs.title
```
I think that this is not correct, even if it works for this case. I believe `.Support("3GB")` shouldn't be equal to `.Support("13GB")`, even when they'll share the same `title`.
You can [compare values with pattern matching](http://www.jessesquires.com/swift-enumerations-and-equatable/), although I think you'll have to cover every case.

frosty replied:
Hmm... the reason I added `Equatable` was simply so that I could find a plan's matching feature compared to the list of all features, for use here: https://github.com/wordpress-mobile/WordPress-iOS/blob/feature/plan-details/WordPress/Classes/ViewRelated/Plans/PlanDetailViewController.swift#L80

This wouldn't work if it also checked on the associated values etc too. I'm happy to restructure the features and do this differently if we can think of a better way though!

koke commented on WordPress/Classes/ViewRelated/Plans/PlanDetailViewController.swift:
```diff
@@ -0,0 +1,202 @@
+    
+    private func configureImmuTable() {
+        viewModel = ImmuTable(sections:
+            [ ImmuTableSection(rows: PlanFeature.allFeatures.map { feature in
```
One of the motivations behind ImmuTable was to be able to define the table contents in a more declarative way. In other words, when you see the ImmuTable being created, it should be obvious which rows it will display. It's a bit more verbose than this solution, but I think the result is more readable.

frosty replied:
As we've discussed, I'll leave this for now as we may be restructuring the PlanFeatures setup in the near future.

koke commented:
The 'WEB ONLY' label is slightly misaligned with respect to the feature title:

![screen shot 2016-02-05 at 16 44 46](https://cloud.githubusercontent.com/assets/8739/12850980/cf9d0968-cc27-11e5-8149-e8d4fa66b740.png)

koke commented on WordPress/Classes/ViewRelated/Plans/PurchaseButton.swift:
```diff
@@ -0,0 +1,52 @@
+        layer.cornerRadius = cornerRadius
+        layer.borderWidth = borderWidth
+        layer.borderColor = tintColor.CGColor
+        layer.borderColor = tintColor.CGColor
```
Duplicate line :wink:

frosty replied:
⌘-C, ⌘-V... Thanks for spotting this!

frosty commented:
I've fixed the WEB ONLY alignment, removed the duplicate line, and replaced the custom cells in `PlanDetailViewController` with standard `WPTableViewCellValue1` cells. I've also added some unit tests to cover the various values that the cells can display.

@koke Would you mind checking over these changes please?

koke commented on WordPress/Classes/ViewRelated/Plans/PurchaseButton.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+import MRProgress
```
I just noticed this. It's not used, right?

frosty replied:
Good spot. No, it's not currently. I was experimenting with using it to display a loading state for the button when a purchase is being made (like the App Store). Doesn't belong on this branch!

koke commented:
:shipit:
